### PR TITLE
Add Codable support to DynamoDB

### PR DESCRIPTION
**Pull request by adam-fowler**
Use Codable to encode/decode [String: DynamoDB.AttributeValue] dictionaries.
Add extensions to DynamoDB for putIem, getItem, query, scan and updateItem that use the encoder and decoder.

### CODE REVIEW

**Comment by Andrea-Scuderi** [comment]
It's really nice! I love it!
It simplify the client code a lot!
I fixed my example:
- before: https://github.com/swift-sprinter/aws-serverless-swift-api-template.git
- after: https://github.com/swift-sprinter/aws-serverless-swift-api-template.git

`scanCodable` is missing? Is it intentional?

**Comment by adam-fowler** [comment]
> `scanCodable` is missing? Is it intentional?

@Andrea-Scuderi  Its good to hear you like it. I don't have a great deal of experience with DynamoDB so wasn't totally sure what I should support while still trying to keep everything fairly unopinionated. I can add a `scanCodable`. Would like to add some form of update but that might need a bit more work.

**Comment by adam-fowler** [comment]
@Andrea-Scuderi added scan and update

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            // construct expression attribute name and value arrays from name attribute value map
+            let expressionAttributeNames: [String: String] = .init( item.keys.map { ("#\($0)", $0) }) { first,_ in return first }
+            let expressionAttributeValues: [String: AttributeValue] = .init( item.map { (":\($0.key)", $0.value) }) { first,_ in return first }
+            // construct update expression, if one if not already supplied
+            let updateExpression: String
+            if let inputUpdateExpression = input.updateExpression {
+                updateExpression = inputUpdateExpression
+            } else {
+                let expressions = item.keys.map {"#\($0) = :\($0)"}
+                updateExpression = "SET \(expressions.joined(separator: ","))"
```
Comment (Andrea-Scuderi) [inline Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:71]: This is quite opinionated.

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        do {
+            var item = try DynamoDBEncoder().encode(input.updateItem)
+            // extract key from input object
+            var key: [String: AttributeValue] = [:]
+            input.key.forEach {
+                key[$0] = item[$0]!
+                item[$0] = nil
+            }
+            // construct expression attribute name and value arrays from name attribute value map
+            let expressionAttributeNames: [String: String] = .init( item.keys.map { ("#\($0)", $0) }) { first,_ in return first }
```
Comment (Andrea-Scuderi) [inline Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:63]: `expressionAttributeNames` could be the default property of UpdateItemCodableInput

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            let expressionAttributeValues: [String: AttributeValue] = .init( item.map { (":\($0.key)", $0.value) }) { first,_ in return first }
```
Comment (Andrea-Scuderi) [inline Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:64]: `expressionAttributeValues` could be the default property of UpdateItemCodableInput

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:71]: If the user supplies no `updateExpression` it is set to `SET...`. So the user can override this

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:64]: see above

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:63]: What would the user setup here apart from the variables taken from the update object?

Comment (adam-fowler) [reply Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift]: I have provided the option to supply your own `expressionAttributeNames` but this is only of any use when the user is supplying their own `updateExpression`.

File: Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/AWSSDKSwift/Extensions/DynamoDB/DynamoDB+Codable.swift:70]: Not sure why you would want this, kinda removes the requirement for the supplied Codable `updateItem`

**Comment by adam-fowler** [comment]
Thinking of removing the word `Codable` from the function names. Function parameters ensure all functions have a different signature from the non-codable versions. It makes the functions slightly more discoverable with code complete. 

Thoughts?

**Pull request merged**